### PR TITLE
Topic remove preprocessor struct generator

### PR DESCRIPTION
**Pull request by psychocoderHPC**
Motivation: The generator is hard to maintain and the usage is not easy
either. With modern C++ and constexpr there is no need for the preprocessor
magic anymore.

- remove the struct generator
- remove the usage of the preprocessor struct generator to create parameters for pre-defined density profiles.

- [x] rebase against #4748

## Note

This PR will close #4748 automatically after it is merged. #4748 produces CI issues those will be gone because this PR is removing the struct generator, therefore this PR must be merged before  #4748.

### CODE REVIEW

**Comment by psychocoderHPC** [comment]
@PrometheusPi Could you have a look into this PR, this was menationed in the dev meeting today

File: include/picongpu/param/particle.param [hunk 1]
```diff
@@ -56,16 +56,13 @@ namespace picongpu
...
-            CONST_VECTOR(float_X, 3, DriftParam_direction, 1.0, 0.0, 0.0);
-
             /** Define Lorentz factor of initial particle drift. */
             struct DriftParam
             {
                 static constexpr float_64 gamma = 1.0;
-                const DriftParam_direction_t direction;
+                /** Define initial particle drift direction vector.
+                 */
+                static constexpr auto driftDirection = float3_X(1.0, 0.0, 0.0);
```
Comment (PrometheusPi) [inline include/picongpu/param/particle.param:65]: in the old code the variable was named `direction` not `driftDirection`.

Comment (PrometheusPi) [reply include/picongpu/param/particle.param:65]: This has also been renamed in `include/picongpu/particles/manipulators/unary/Drift.hpp` :+1:

File: include/pmacc/math/vector/Vector.hpp [hunk 2]
```diff
@@ -23,7 +23,7 @@
 #pragma once
 
 #include "pmacc/algorithms/math.hpp"
-#include "pmacc/math/vector/navigator/StandardNavigator.hpp"
```
Comment (PrometheusPi) [inline include/pmacc/math/vector/Vector.hpp:26]: has `StandardNavigator.hpp` been removed?

File: share/picongpu/examples/TransitionRadiation/include/picongpu/param/density.param [hunk 3]
```diff
@@ -44,30 +41,33 @@ namespace picongpu
...
              */
-            (PMACC_C_VALUE(uint32_t, vacuumCellsY, 0))
+            static constexpr uint32_t vacuumCellsY = 0u;
 
             /** The central position of the density distribution
              *  unit: meter
              */
-            (PMACC_C_VECTOR_DIM(float_64, simDim, center_SI, 128 * 0.16e-6 / 2.0, 0.912e-5, 128 * 0.16e-6 / 2.0))
+            static constexpr floatD_64 center_SI
+                = float3_64(128 * 0.16e-6 / 2.0, 0.912e-5, 128 * 0.16e-6 / 2.0).shrink<simDim>();
```
Comment (PrometheusPi) [inline share/picongpu/examples/TransitionRadiation/include/picongpu/param/density.param:65]: not relevant for this PR: 
@finnolec why is the resolution in y so low?

File: share/picongpu/pypicongpu/template/include/picongpu/param/particle.param.mustache [hunk 4]
```diff
@@ -92,7 +87,11 @@ namespace picongpu
                             *  unit: none
                             */
                             static constexpr float_64 gamma = {{{gamma}}};
-                            const drift_{{{species.typename}}}_direction_t direction;
+                            {{#direction_normalized}}
```
Comment (PrometheusPi) [inline share/picongpu/pypicongpu/template/include/picongpu/param/particle.param.mustache:90]: @BrianMarre could you have a look at this mustache code?

**Review by PrometheusPi — COMMENTED** [review]
Looks good to me. 
Just a few questions (not necessarily to @psychocoderHPC) that do not prevent merging.

**Review by PrometheusPi — APPROVED** [review]
Did not want to comment but to approve.

**Pull request merged**